File: solutions/cpp/Rebuild_BST_postorder.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <limits>
#include <memory>
#include <string>
#include <vector>

#include "./BST_prototype.h"

using std::cout;
using std::endl;
using std::numeric_limits;
using std::unique_ptr;
using std::vector;

BSTNode<int>* rebuild_BST_postorder_helper(
    const vector<int> &postorder, int s, int e);

// @include
// Given a postorder traversal of a BST, return its root.
BSTNode<int>* rebuild_BST_from_postorder(const vector<int>& postorder) {
  return rebuild_BST_postorder_helper(postorder, 0, postorder.size());
}

// Build a BST based on postorder[s : e - 1], return its root.
BSTNode<int>* rebuild_BST_postorder_helper(
    const vector<int> &postorder, int s, int e) {
  if (s < e) {
    int x = s;
    while (x < e && postorder[x] < postorder[e - 1]) {
      ++x;
    }
    return new BSTNode<int>{
        postorder[e - 1],
        unique_ptr<BSTNode<int>>(
          rebuild_BST_postorder_helper(postorder, s, x)),
        unique_ptr<BSTNode<int>>(
          rebuild_BST_postorder_helper(postorder, x, e - 1))
    };
  }
  return nullptr;
}
// @exclude

template <typename T>
void check_ans(const unique_ptr<BSTNode<T>>& n, const T& pre) {
  if (n) {
    check_ans(n->left, pre);
    assert(pre <= n->data);
    cout << n->data << endl;
    check_ans(n->right, n->data);
  }
}

int main(int argc, char *argv[]) {
  //      1
  //        2
  //          3
  //            4
  //              5
  //                6
  // should output 1, 2, 3, 4, 5, 6
  // postorder [6, 5, 4, 3, 2, 1]
  vector<int> postorder;
  postorder.emplace_back(6);
  postorder.emplace_back(5);
  postorder.emplace_back(4);
  postorder.emplace_back(3);
  postorder.emplace_back(2);
  postorder.emplace_back(1);
  unique_ptr<BSTNode<int>> root(rebuild_BST_from_postorder(postorder));
  check_ans(root, numeric_limits<int>::min());
  return 0;
}

File: static/solutions/java/src/main/java/com/epi/RebuildBSTPostorder.java
package com.epi;

import com.epi.BinarySearchTreePrototypeTemplate.BSTNode;

import java.util.ArrayList;
import java.util.List;

public class RebuildBSTPostorder {
  // @include
  // Given a postorder traversal of a BST, return its root.
  public static BSTNode<Integer> rebuildBSTFromPostorder(
      List<Integer> postorder) {
    return rebuildBSTPostorderHelper(postorder, 0, postorder.size());
  }

  // Build a BST based on postorder.subList(s, e), return its root.
  private static BSTNode<Integer> rebuildBSTPostorderHelper(
      List<Integer> postorder, int s, int e) {
    if (s < e) {
      int x = s;
      while (x < e && postorder.get(x) < postorder.get(e - 1)) {
        ++x;
      }
      return new BSTNode<>(postorder.get(e - 1),
                           rebuildBSTPostorderHelper(postorder, s, x),
                           rebuildBSTPostorderHelper(postorder, x, e - 1));
    }
    return null;
  }
  // @exclude

  private static void checkAns(BSTNode<Integer> n, Integer pre) {
    if (n != null) {
      checkAns(n.left, pre);
      assert(Integer.compare(pre, n.data) <= 0);
      System.out.println(n.data);
      checkAns(n.right, n.data);
    }
  }

  public static void main(String[] args) {
    // 1
    // 2
    // 3
    // 4
    // 5
    // 6
    // should output 1, 2, 3, 4, 5, 6
    // postorder [6, 5, 4, 3, 2, 1]
    List<Integer> postorder = new ArrayList<>();
    postorder.add(6);
    postorder.add(5);
    postorder.add(4);
    postorder.add(3);
    postorder.add(2);
    postorder.add(1);
    BSTNode<Integer> root = rebuildBSTFromPostorder(postorder);
    checkAns(root, Integer.MIN_VALUE);
  }
}
